### ISSUE TITLE
Expanding a subprocess overlaps previous content

### ISSUE THREAD

**Issue by volkergersabeck** [camunda/camunda-modeler#1244]
When a user adds a task
And changes this task into an expanded subprocess
Then the subprocess overlaps previous elements

Expected behaviour: if the subprocess does not yet have following content, then expand the subprocess only to the right, such that it does not overlap the previous content

![1244 switch task to subprocess](https://user-images.githubusercontent.com/232571/52943874-62e36680-336e-11e9-88e5-a8aa4e69f1ce.gif)

**Comment by nikku** [comment]
Nice little improvement.

**Comment by falko** [comment]
In other situations, an automatic application of the Space Tool might help.

**Comment by philippfromme** [comment]
We could even go as far as checking for available space in all directions and expanding based on that information.

**Comment by falko** [comment]
Any improvement in this area would be helpful.

**Comment by falko** [comment]
And let's very be clear here: This is how it always looks when one creates a Sub-Process because our users are conditioned to avoid the palette and use the context pad.

**Comment by gustavjf** [comment]
@philippfromme

Right, we have an [issue](https://github.com/camunda/camunda-modeler/issues/1243) for the scenario where there is content after the expanded subprocess. In the case of content above and below being overlapped, I think that would need further discussion on what the expected result should be. In this ticket, I'll focus on the scenario where there is only content before, and not above/below/after.

@falko

Just to make sure I understand what you mean by "one creates", do you mean the same as when "one expands", as shown in the screen capture? If not, could you explain how this issue is manifested when creating an expanded subprocess? I understand creating and expanding to be two different scenarios unless I'm mistaken.

**Comment by falko** [comment]
The default way to create any activity in the "normal flow" is to open 
the context pad on the current element, create an abstract task and use 
the wrench to morph it into the desired task type. Users do the same 
with an embedded sub-process, except that in this case the activity 
becomes bigger and currently breaks the layout.

There is also a sub-process in the palette but that is mostly used for 
event sub-processes that are outside the "normal flow", i.e. not 
connected with sequence flows. For elements in the "normal flow", users 
are rewarded for using the context pad with the automatic positioning 
and connecting of the new element.

In my opinion, "expanding" and "collapsing" of sub-processes is not 
relevant for practical use, because the contents of collapsed 
sub-process will be hidden in all visualizations rendered in any other 
Camunda tool, e.g. heatmaps in Cockpit and Optimize. Supporting this 
properly across all tools would be a much bigger discussion.

However, creating an (expanded) embedded sub-process is an everyday 
operation and should be a smooth experience for users.

**Comment by gustavjf** [comment]
Got it, thanks for the clarification.

**Comment by nikku** [comment]
We moved this issue back to the backlog as the implementation turned out way more complicated than expected.

**Comment by nikku** [comment]
Work in progress on upstream feature branch: https://github.com/bpmn-io/bpmn-js/compare/1244-expand-subprocess-right.

**Comment by JacobiCamunda** [comment]
I stumbled across this today. Here is what it currently looks like.
[screen-recorder-wed-jan-17-2024-13-57-48.webm](https://github.com/camunda/camunda-modeler/assets/108456325/aee01186-afdf-4e0d-a553-ab57dd38a33e)

**Comment by falko** [comment]
The same happened to me in recent customer meetings and it's always pretty embarrassing. Would it be an option to have a subprocess directly in the context pad so that the layout assist can place it further to the right immediately? Alternatively, the morph would have to perform some layout assist, i.e. if the target shape is bigger that the current shape, move the shape to the right if there is space or make space for it.

**Comment by nikku** [comment]
CC @crobbins215. A good example where we're not done yet and cause users in rather common refactoring flows during refinement harm.

A solution is, unfortunately, [not an easy pick](https://github.com/camunda/camunda-modeler/issues/1244#issuecomment-496835383), as it would require a [flow local space tool](https://github.com/camunda/camunda-modeler/issues/876).

We could re-assess if there is easy picks to improve the situation.

**Comment by crobbins215** [comment]
We had this exact issue when we added "groups" to Flow, we didn't get around to the solution which would be something like https://github.com/camunda/camunda-modeler/issues/876.

**Comment by falko** [comment]
How about we just do a partial solution and assume that there is already space to the right, e.g. because we are at the end of a sequence or have already created space. With that assumption we could simply place the Sub-Process further right, i.e. aligned to the left of the morphed task and not centered on it. Maybe that is a quick win that solves 80%.

**Comment by nikku** [comment]
We tried the partial solution. It was not an easy pick at the time. 

If we optimized for the partial solution, why not use append anything to append a sub-process right away using create/append anything?

![capture 4dWlYy_optimized](https://github.com/camunda/camunda-modeler/assets/58601/1c72c437-2629-4bda-ac8a-ed2e44eba8e1)

**Comment by barmac** [comment]
[This is still a problem](https://github.com/camunda/camunda-modeler/issues/5266) even with append-anything. I'd like us to revisit it.

CC @Balazs-PM

**Comment by nikku** [comment]
Especially relevant in the context of agentic, which we're pursuing in the future.

CC @YanaSegal 😉

**Comment by YanaSegal** [comment]
This is a big issue now with the agent - see more description here https://github.com/camunda/camunda-modeler/issues/5266

(also it's pretty impressive this issue exists since 2019 😅)

**Comment by nikku** [comment]
> (also it's pretty impressive this issue exists since 2019 😅)

It is also easy to explain. It is a hard problem to solve, that has to date not been prioritized.

**Comment by bojtospeter** [comment]
Hey team, with the communication to use AI Agent processes, it might be confusing to resize the agent itself which then "hides" your bpmn elements. I don't know how big of a blocker it is, but wondering if subprocesses should have a lower "Z" index or something to not hide other elements. I like the gimmick, that the subprocess is semi-transparent, showing that there's something behind. :)
(I guess it's more complicated than this, otherwise someone would have already solved it in 5 minutes.)

https://github.com/user-attachments/assets/f67c170e-67f4-41ea-89f9-e639825de976

**Comment by barmac** [comment]
Would you expect that by expansion the elements are moved inside of the subprocess?

**Comment by YanaSegal** [comment]
> but wondering if subprocesses should have a lower "Z" index or something to not hide other elements.
Users may think that the elements are part of the sub-process. 

optional - auto space when resizing, or if we add them to the sub process as Maciej asked above, then we need to visualize this before. 

What is weird is that the element is hidden, but the doc icon has a higher "Z" index - also confusing...

**Comment by falko** [comment]
Hidden elements should be prevented or at least warned about by the UX, e.g. by rendering them in bright red and not just semi-transparent. I can't think of any situation, in which they are desirable. Auto Space would be awesome but I'm sure there can be very difficult cases plus it should only kick in if there is not enough space, e.g. when the sub-process starts to overlap with other element. Automatically "sucking in" overlapped elements when resizing feels like a step to far unless it is done with a special tool like the lasso-style tool for creating a sub-process out of already existing element that we dream about elsewhere.

**Comment by nikku** [comment]
👍 💯 to https://github.com/camunda/camunda-modeler/issues/1244#issuecomment-3645488216.

▶️ Boxing matters, what visually "is contained" _must_ be also semantically contained - if a start event looks like it is _in the sub-process_, then it _must be inside of the sub-process_.
🤔 As per our [design principles](https://github.com/bpmn-io/design-principles) we don't want to do magic. Already today you are reasonably well supported when you intend to move elements into a sub-process, and [wrapping into sub-processes](https://github.com/camunda/camunda-modeler/issues/177) could be an additional action.

What we need is incremental (non-destructive) layouting to advance "recursive expansion" of the model, supporting resize without overlap, without surprises.

---

![Image](https://github.com/user-attachments/assets/a433b9bb-90ff-4aa4-ba99-5a123fcadc05)
